File: Data Structures & Algorithms/sqrtx/submission-0.py
class Solution:
    def mySqrt(self, x: int) -> int:

        if x < 2:
            return x

        left = 1
        right = x
        answer = 0

        while left <= right:
            mid = (left + right) // 2

            if mid * mid == x:
                return mid

            elif mid * mid < x:
                # mid is a possible answer
                answer = mid
                left = mid + 1

            else:
                # mid is too large
                right = mid - 1

        return answer

File: Data Structures & Algorithms/eating-bananas/submission-0.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1;
        int right = 0;

        // Maximum pile is the upper bound for k
        for (int pile : piles) {
            right = Math.max(right, pile);
        }

        int answer = right;

        while (left <= right) {
            int k = left + (right - left) / 2;

            long hours = 0;

            for (int pile : piles) {
                // Ceiling(pile / k)
                hours += (pile + (long) k - 1) / k;
            }

            if (hours <= h) {
                // k works, try a smaller speed
                answer = k;
                right = k - 1;
            } else {
                // k is too slow
                left = k + 1;
            }
        }

        return answer;
    }
}